swarm/pss/pss.go

```diff
@@ -241,91 +317,283 @@ func (self *Pss) checkFwdCache(addr []byte, digest pssDigest) bool {
+	}
+	// if the send fails it means this public key is not registered for this particular address AND topic
+	err = self.SendAsym(to, *topic, recvkeybytes)
+	log.Debug("recvkeybytes", "bytes", recvkeybytes, "recvkey", recvkey)
+	if err != nil {
+		return "", fmt.Errorf("Send symkey failed: %v", err)
+	}
+	return recvkeyid, nil
+}
+
+// Prepares a msg for sending with symmetric encryption
 //
-// This method is payload agnostic, and will accept any arbitrary byte slice as the payload for a message.
+// this can only succeed if there exist unexpired symmetric keys both for incoming and outgoing traffic. This will be the state after a asymmetric exchange of symmetric keys (handshake)
+func (self *Pss) SendSym(to []byte, topic whisper.TopicType, msg []byte) error {
+	var potaddr pot.Address
+	copy(potaddr[:], to)
+	// isSecured also checks whether the first dimension of the map is populated
+	if !self.isSecured(potaddr, topic) {
+		return fmt.Errorf("missing complete handshake")
+	}
+	psp := self.peerPool[potaddr][topic]
+	symkey, err := self.w.GetSymKey(psp.sendsymkey)
+	if err != nil {
+		return fmt.Errorf("missing valid symkey %s: %v", psp.sendsymkey, err)
+	}
+	return self.send(to, topic, msg, nil, symkey)
+}
+
+// Prepares a msg for sending with asymmetric encryption
 //
-// It generates an envelope for the specified recipient and topic, and wraps the message payload in it.
-func (self *Pss) SendRaw(to []byte, topic Topic, msg []byte) error {
-	sender := self.Overlay.BaseAddr()
-	pssenv := NewEnvelope(sender, topic, msg)
+// Asymmetric send can be used to exchange symmetric keys (handshake)
+func (self *Pss) SendAsym(to []byte, topic whisper.TopicType, msg []byte) error {
+	var potaddr pot.Address
+	copy(potaddr[:], to)
+	topicmap := self.peerPool[potaddr]
+	if topicmap == nil {
+		return fmt.Errorf("No public key for address %x", to)
+	}
+	psp := self.peerPool[potaddr][topic]
+	topubkey := psp.pubkey
+	return self.send(to, topic, msg, topubkey, nil)
+}
+
+// Sends a pss message
+//
+// The method itself is payload agnostic, and will accept any arbitrary byte slice as the payload for a message.//
+// It generates an whisper envelope for the specified recipient and topic, and wraps the message payload in it.
+func (self *Pss) send(to []byte, topic whisper.TopicType, msg []byte, pubkey *ecdsa.PublicKey, symkey []byte) error {
+	if pubkey != nil && symkey != nil {
+		return fmt.Errorf("Can only specify one of symkey and pubkey for send")
+	} else if pubkey == nil && symkey == nil {
+		return fmt.Errorf("No keys set, refusing to send unencrypted pss")
+	}
+	wparams := &whisper.MessageParams{
+		TTL:      DefaultTTL,
+		Src:      self.privatekey,
+		Dst:      pubkey, // if this is set, the message is considered asymmetric, if not it is considered symmetric
+		KeySym:   symkey,
+		Topic:    topic,
+		WorkTime: defaultWhisperWorkTime,
+		PoW:      defaultWhisperPoW,
+		Payload:  msg,
+		Padding:  []byte("1234567890abcdef"),
```

Comment:
I think the point is to make packet analysis based on content LENGHT trickier, and doesn't have too much to do with the content itself. But I would assume that it's better to choose a good random slice (which has nothing to do with the data itself) rather than a static value like 0's. As mentioned, I think I would like some input from higher expetise on this exact theme.